Lib/shelve.py

```diff
@@ -110,19 +124,16 @@ def __getitem__(self, key):
         try:
             value = self.cache[key]
         except KeyError:
-            f = BytesIO(self.dict[key.encode(self.keyencoding)])
-            value = Unpickler(f).load()
+            f = self.dict[key.encode(self.keyencoding)]
+            value = self.deserializer(f)
             if self.writeback:
                 self.cache[key] = value
         return value
 
     def __setitem__(self, key, value):
         if self.writeback:
             self.cache[key] = value
-        f = BytesIO()
-        p = Pickler(f, self._protocol)
-        p.dump(value)
-        self.dict[key.encode(self.keyencoding)] = f.getvalue()
+        self.dict[key.encode(self.keyencoding)] = self.serializer(value, self._protocol)
```

Comment:
The second argument for the serializer needs to be described  in the docs. Or can we refactor such that the serializer only takes a single argument?

Reply:
Second argument for the serializer described in the docs.